File: c++/0046. Longest Substring without Repeating Characters.cpp
// time = O(n)
// space = O(1)

class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        vector<int> char_freq(255, -1);
        
        int curr = 0;
        int longest_substrong = 0;
        for (int i = 0; i < s.size(); i++) {
            if (char_freq[s[i]] < i - curr) {
                curr++;
                char_freq[s[i]] = i;
                if (curr > longest_substrong) longest_substrong = curr;
            } else {
                curr = i - char_freq[s[i]];
                char_freq[s[i]] = i;
            }
        }
        return longest_substrong;
    }
};

File: 0046_longest_substring_without_repeating_characters.py
# time: O(n)
# space: O(n)
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        letter_pos = {}
        f_ptr, longest = 0, 0
        for idx, char in enumerate(s):
            if char in letter_pos and letter_pos[char] >= f_ptr:
                f_ptr = letter_pos[char] + 1
            else:
                longest = max(longest, idx-f_ptr+1)
            letter_pos[char] = idx
        return longest